### ISSUE TITLE
Dialog with asynchronously loaded content is removed during page load

### ISSUE THREAD

**Issue by MikeParkin** [orocrm/platform#240]
Inside my .html.twig file I create a new dialog box that shows a toolbar.

The toolbar has items in to be used on this page, so should be loaded as soon as possible (rather than opening on a button press, which might be a more typical use case).

Here is the code:

```javascript
require(['oro/dialog-widget', 'routing'], function(Dialog, routing){
   var dialog = new Dialog({
       url: routing.generate('defacto_toolbar'),
       title: ('Content Toolbar - stand alone'),
       dialogOptions: {
           width: 260,
           height: 500
       },
       stateEnabled: false
   });
    dialog.render();
});
```
The content for my dialog is loaded asynchronously from the server (because the contents are dynamic). When my page loads I get the following error:

> Uncaught TypeError: Cannot assign to read only property 'loading' of [object Object]

I've traced this down to the Widget-Manager class removing widgets without a Dom element. I believe this was added to help fix memory leaks:

```javascript
/**
 * Reset manager to initial state.
 */
resetWidgets: function() {
    _.each(this.widgets, function (widget) {
        // if widget is detached from DOM, remove this widget
        if (!$.contains(document.documentElement, widget.el)) {
            widget.remove();
        }
    });
},
```
Unfortunately it seems my widget is removed from the page, because it does not have a widget.el at this point (the contents are still being loaded via Ajax). When the widget is removed it is disposed, which causes all of it's properties to be set to read-only.

Is there a suggested work around to this? Am I forced to load my widget at a later point, or is my use-case valid and you will implement a fix? I am guessing that other widget types could experience the same issue, if they load their content asynchronously at the start of the request.

**Comment by MikeParkin** [comment]
For my use-case, I can simply not allow the removal of the dialog widget if the content is still loading.

I did this with a local override, as it only affects my dialogs.

This feels like a simple, but not perfect fix. Please let me know your thoughts.

```javascript
define(['oro/dialog-widget'], function (DialogWidgetAbstract) {
    'use strict';

    /**
     * Do not remove the widget if it is still loading it's contents
     * 
     * @see https://github.com/orocrm/platform/issues/240
     **/
    var DialogWidget = DialogWidgetAbstract.extend({
        remove: function(){
            if (! this.loading){
                DialogWidget.__super__.remove.call(this);
            }
        }
    });

    return DialogWidget;
});
```

Core fix for this could be:

```javascript
        /**
         * Removes dialog widget
         */
        remove: function() {
            if (this.loading) {
                return;
            }
            if (this.widget) {
                // There's widget, close it before remove.
                // Close handler will invoke dispose method,
                // where remove method will be called again
                this.widget.dialog('close');
            } else {
                DialogWidget.__super__.remove.call(this);
            }
        },
```

**Comment by Grygir** [comment]
Yes, this is expected behavior. If a widget's element is detached from DOM, widget manager destroys this widget on `pageAfterChange` event. If to add exception for loading widget (as you proposed) what would be a natural way to dispose this widget?

For your case, I would recommend you to add the dialog to DOM manually.
```javascript
require(['jquery', 'oro/dialog-widget', 'routing'], function($, Dialog, routing){
    var dialog = new Dialog({
        url: routing.generate('oro_entityconfig_widget_info', {id: 1}),
        title: ('Content Toolbar - stand alone'),
        dialogOptions: {
            width: 260,
            height: 500
        },
        stateEnabled: false
    });
    $(document.body).append(dialog.$el);
    dialog.render();
});

**Comment by MikeParkin** [comment]
Hi @Grygir - Thanks for the feedback and work around, this does work for me.

To answer your question - I could not think of a time that I would like to remove a dialog widget whilst it was still loading - can you?

It is slightly hidden that you need to add Dialog.$el to the DOM manually, but only if you load dialog with URL before pageAfterChange event?

What are your thoughts on the manager checking this state? That way dispose would still work but I would not need to add Dialog to the DOM manually for Ajax Dialogs?

Any scenarios I have missed here?

```javascript
/**
 * Reset manager to initial state.
 */
resetWidgets: function() {
    _.each(this.widgets, function (widget) {
        // if widget is detached from DOM and not loading, remove this widget
        if (!$.contains(document.documentElement, widget.el) && !widget.loading) {
            widget.remove();
        }
    });
},
```

**Comment by Grygir** [comment]
Hi @MikeParkin 

Loading is not sufficient sign that widget is still actual. There might some error have happened and flag loading is not removed, we have to be able to clean up such widget on page change action. I see the problem that pointing out. Think we will add a solution in to coming release.

Thanks for your question.

**Comment by artml** [comment]
Fixed in 1.7
